transactions concept doc

gzhao408 opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/develop/CONTRIBUTING.md#pr-targeting))

- [ ] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry: `clog add [section] [stanza] [message]`
- [ ] rereviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

gzhao408 commented:
#4705

gamarin2 commented on docs/concepts/transactions.md:
```diff
@@ -0,0 +1,51 @@
+
+## Synopsis
+
+This doc describes how transactions are defined.
```
no abbreviation please!

gamarin2 commented on docs/concepts/transactions.md:
```diff
@@ -0,0 +1,51 @@
+
+## Transactions
+
+[**Transactions**](https://github.com/cosmos/cosmos-sdk/blob/97d10210beb55ad4bd6722f7186a80bf7cb140e2/types/tx_msg.go#L36-L43)  trigger state changes. They are comprised of metadata and [messages](./modules.md#messages) that instruct modules to perform various actions.
```
`Messages` do not instruct modules, they are handled by module's handlers in order to trigger state transitions

gamarin2 commented on docs/concepts/transactions.md:
```diff
@@ -0,0 +1,51 @@
+
+[**Transactions**](https://github.com/cosmos/cosmos-sdk/blob/97d10210beb55ad4bd6722f7186a80bf7cb140e2/types/tx_msg.go#L36-L43)  trigger state changes. They are comprised of metadata and [messages](./modules.md#messages) that instruct modules to perform various actions.
+
+When users interact with an application and make state changes (e.g. sending coins or changing stored values), transactions are created. Transactions must be signed by the appropriate account holders as specified in the transaction definitions, then broadcast to the network. These transactions must then be included in a block, validated, and approved by the network through the consensus process.  ...
```
want to interract

they create transactions

gamarin2 replied:
what is account holder? 

broadcasted

gamarin2 replied:
each message has a signature, not the tx

gamarin2 commented on docs/concepts/transactions.md:
```diff
@@ -0,0 +1,51 @@
+
+## Transaction Definition
+
+Transactions are defined by module developers and application developers. They must implement the [`Tx`](https://github.com/cosmos/cosmos-sdk/blob/73700df8c39d1fe6c3d3a1a635ac03d4bacecf55/types/tx_msg.go#L34-L41) interface and include an encoder and decoder.
```
modules developers

gamarin2 replied:
Add transition sentence like "Here are the important methods of the `Tx` interface"

gamarin2 commented on docs/concepts/transactions.md:
```diff
@@ -0,0 +1,51 @@
+Transactions are defined by module developers and application developers. They must implement the [`Tx`](https://github.com/cosmos/cosmos-sdk/blob/73700df8c39d1fe6c3d3a1a635ac03d4bacecf55/types/tx_msg.go#L34-L41) interface and include an encoder and decoder.
+
+* **GetMsgs:** unwraps the transaction and returns a list of its message(s) - one transaction may have one or multiple messages.
+* **ValidateBasic:** includes lightweight, stateless checks on the validity of a transaction.
```
Stateless/Stateful should be in italic

gamarin2 replied:
explain difference with `ValidateBasic` for message in practice

gamarin2 commented on docs/concepts/transactions.md:
```diff
@@ -0,0 +1,51 @@
+
+* **GetMsgs:** unwraps the transaction and returns a list of its message(s) - one transaction may have one or multiple messages.
+* **ValidateBasic:** includes lightweight, stateless checks on the validity of a transaction.
+* **TxEncoder:** Nodes running the consensus engine (e.g. Tendermint Core) process transactions in the `[]byte` form, so transactions are always encoded before being handled by Tendermint processes. The Cosmos SDK allows developers to specify any deterministic encoding format for their applications; the default is [`Amino`](./amino.md).
```
process -> gossip

This paragraph needs to be reworded. Explain that the consensus engine is responsible for ordering and gossiping transaction `[]bytes`. It only knows txs in their encoded form (i.e. `[]byte`), to maintain agnosticism wrt the application and facilitate gossiping.

gamarin2 commented on docs/concepts/transactions.md:
```diff
@@ -0,0 +1,51 @@
+* **GetMsgs:** unwraps the transaction and returns a list of its message(s) - one transaction may have one or multiple messages.
+* **ValidateBasic:** includes lightweight, stateless checks on the validity of a transaction.
+* **TxEncoder:** Nodes running the consensus engine (e.g. Tendermint Core) process transactions in the `[]byte` form, so transactions are always encoded before being handled by Tendermint processes. The Cosmos SDK allows developers to specify any deterministic encoding format for their applications; the default is [`Amino`](./amino.md).
+* **TxDecoder:** [ABCI](https://tendermint.com/docs/spec/abci/) calls such as `CheckTx` and `DeliverTx` can be used to process transaction data at the application layer; they are passed in as `txBytes []byte` an first unmarshaled (decoded) using `TxDecoder` before any logic is applied.
```
When the application receives transactions from the underlying consensus engine via ABCI messages like `CheckTx` and `DeliverTx`, they are in their standard encode `[]byte` format. Methods handling these ABCI messages like `CheckTx()` and `DeliverTx()` will first decode the transaction bytes using `TxDecoder` before applying any logic. T

gamarin2 replied:
+ link to `./baseapp.md#checktx` and delivertx as well

gamarin2 commented on docs/concepts/transactions.md:
```diff
@@ -0,0 +1,51 @@
+
+## Messages
+
+**Messages** describe possible actions within a module. Developers define the specific messages for each application module by implementing the [`Msg`](https://github.com/cosmos/cosmos-sdk/blob/97d10210beb55ad4bd6722f7186a80bf7cb140e2/types/tx_msg.go#L10-L31) interface, and also define a [`Handler`](./handlers.md) that performs the specified actions. All the messages in a module are typically def ...
```
`Messages` are objects created by the end-user in order to trigger state transitions.

gamarin2 replied:
a `handler` to process them

gamarin2 replied:
not necessarily defined in a `msgs.go` file

gamarin2 replied:
`handler` doc is in `../building-modules/handler.md` now

gamarin2 commented on docs/concepts/transactions.md:
```diff
@@ -0,0 +1,51 @@
+* **Route** returns a string that specifies which module this message is a part of and, thus, which module contains the handler used to implement the actions of this message. Applications that inherit from [baseapp](./baseapp.md) use this function to help deliver transactions by invoking the correct handler(s). For example, `MsgSend` is defined in and handled by the `bank` module; its `Route` fun ...
+* **Type** returns a short, human-readable string that describes what the message does. For example, `MsgSend`'s type is `"send"`.
+* **ValidateBasic** implements a set of stateless sanity checks for the message and returns an `Error`. For example, the `validateBasic` method for `MsgSend`, the message which sends coins from one account to another, checks that both sender and recipient addresses are valid and the amount is nonnegative.
+* **GetSignBytes** returns a `[]byte` representation of the message.
```
Why? in order for the sender to sign the message

gamarin2 commented on docs/concepts/transactions.md:
```diff
@@ -0,0 +1,63 @@
+
+## Transaction Generation
+
+[`Contexts`](https://godoc.org/context) are immutable data structures that can be cloned and updated as it is passed between processes and request handlers. In the process of creating a transaction, two contexts are created: the `CLIContext` and `TxBuilder`. Both are automatically generated and do not need to be defined by the developer, but do require input from the transaction creator (e.g. usi ...
```
First sentence is not right

gamarin2 replied:
Also "Context is an immutable object that contains all information needed to process a request" + give examples of what it contains

gamarin2 replied:
Is the `TxBuilder` a context? Also `TxBuilder` may require more detail, like a codeblock to show the struct and an explanation of the various parameters

gamarin2 commented on docs/concepts/transactions.md:
```diff
@@ -0,0 +1,63 @@
+
+[`Contexts`](https://godoc.org/context) are immutable data structures that can be cloned and updated as it is passed between processes and request handlers. In the process of creating a transaction, two contexts are created: the `CLIContext` and `TxBuilder`. Both are automatically generated and do not need to be defined by the developer, but do require input from the transaction creator (e.g. usi ...
+
+Every message in a transaction must be signed by the addresses specified by `GetSigners`. The signing process is handled by the [`auth`](https://github.com/cosmos/cosmos-sdk/tree/67f6b021180c7ef0bcf25b6597a629aca27766b8/docs/spec/auth) module and automatically performed when the transaction is created, unless the user choses to generate and sign separately. The `TxBuilder` contains the `KeyBase`  ...
```
The signing process must be handled by a module. The most widely used module for that purpose is the `auth` module

gamarin2 commented on docs/concepts/transactions.md:
```diff
@@ -0,0 +1,63 @@
+
+Every message in a transaction must be signed by the addresses specified by `GetSigners`. The signing process is handled by the [`auth`](https://github.com/cosmos/cosmos-sdk/tree/67f6b021180c7ef0bcf25b6597a629aca27766b8/docs/spec/auth) module and automatically performed when the transaction is created, unless the user choses to generate and sign separately. The `TxBuilder` contains the `KeyBase`  ...
+
+## Handlers and Keepers
```
I think mentioning the Handler is sufficient here

fedekunze commented:
is this ready for review? what's the status of this PR?

gzhao408 commented:
> is this ready for review? what's the status of this PR?

@fedekunze waiting for Gautier to take a look again before marking R4R, but yes it's ready to be looked at

fedekunze commented:
@gzhao408 cool I'll mark this as R4R then. Mind resolving the conversations that are already resolved?

gzhao408 commented on docs/core/transactions.md:
```diff
@@ -0,0 +1,84 @@
+
+* **GetMsgs:** unwraps the transaction and returns a list of its message(s) - one transaction may have one or multiple messages.
+* **ValidateBasic:** includes lightweight, [*stateless*](../basics/tx-lifecycle.md#types-of-checks) checks used by ABCI messages [`CheckTx`](../basics/baseapp.md#checktx) and [`DeliverTx`](../basics/baseapp.md#delivertx) to validate transactions. The `StdTx` `ValidateBasic` function checks that its transactions are signed by the correct number of signers and that the fees do not exceed what the u ...
+* **TxEncoder:** Nodes running the consensus engine (e.g. Tendermint Core) are responsible for gossiping transactions and ordering them into blocks, but only handle them in the generic `[]byte` form. Transactions are always marshaled (encoded) before they are relayed to nodes, which compacts them to facilitate gossiping and helps maintain the consensus engine's agnosticism with applications. The  ...
```
thanks :)

gzhao408 commented on docs/core/transactions.md:
```diff
@@ -0,0 +1,84 @@
+
+## Handlers
+
+The final components developers must implement to enable transactions are handlers and keeprs. Each module has a [`Handler`](../building-modules/handler.md) to process all of the module's message types.
```
didn't want to bloat this doc haha, but added a bit of finality to it. I expect when Gautier has finalized the flow of the docs he'll add a "Next, read about ___" section at the end

fedekunze commented on docs/core/transactions.md:
```diff
@@ -0,0 +1,84 @@
+Transactions messages are defined by module developers. The transaction objects themselves are SDK types that implement the [`Tx`](https://github.com/cosmos/cosmos-sdk/blob/73700df8c39d1fe6c3d3a1a635ac03d4bacecf55/types/tx_msg.go#L34-L41) interface and include an encoder and decoder:
+
+* **GetMsgs:** unwraps the transaction and returns a list of its message(s) - one transaction may have one or multiple messages.
+* **ValidateBasic:** includes lightweight, [*stateless*](../basics/tx-lifecycle.md#types-of-checks) checks used by ABCI messages [`CheckTx`](../basics/baseapp.md#checktx) and [`DeliverTx`](../basics/baseapp.md#delivertx) to validate transactions. The `StdTx` `ValidateBasic` function checks that its transactions are signed by the correct number of signers and that the fees do not exceed what the u ...
```
use master for the spec instead of the commit

fedekunze commented on docs/core/transactions.md:
```diff
@@ -0,0 +1,84 @@
+
+The `CLIContext` is initialized using the application's `codec` and data more closely related to the user interaction with the interface, holding data such as the output to the user and the broadcast mode. Read more about `CLIContext` [here](../interfaces/query-lifecycle.md#clicontext).
+
+Every message in a transaction must be signed by the addresses specified by `GetSigners`. The signing process must be handled by a module, and the most widely used one is the [`auth`](https://github.com/cosmos/cosmos-sdk/tree/67f6b021180c7ef0bcf25b6597a629aca27766b8/docs/spec/auth) module. Signing is automatically performed when the transaction is created, unless the user choses to generate and s ...
```
ditto

rigelrozanski commented:
Why dis failing? maybe update to master.. also @hschoenburg have you comments been thoroughly addressed?

gzhao408 commented:
> Why dis failing? maybe update to master.. also @hschoenburg have you comments been thoroughly addressed?

@rigelrozanski just addressed all the comments and rebased

gamarin2 commented on docs/core/transactions.md:
```diff
@@ -0,0 +1,84 @@
+
+## Transaction Definition
+
+Transactions messages are defined by module developers. The transaction objects themselves are SDK types that implement the [`Tx`](https://github.com/cosmos/cosmos-sdk/blob/master/types/tx_msg.go#L34-L41) interface and include an encoder and decoder:
```
The first sentence should be under `GetMsgs`. The section should not start with messages if it's about transactions, cause that's confusing

gamarin2 commented on docs/core/transactions.md:
```diff
@@ -0,0 +1,84 @@
+
+Transactions messages are defined by module developers. The transaction objects themselves are SDK types that implement the [`Tx`](https://github.com/cosmos/cosmos-sdk/blob/master/types/tx_msg.go#L34-L41) interface and include an encoder and decoder:
+
+* **GetMsgs:** unwraps the transaction and returns a list of its message(s) - one transaction may have one or multiple messages.
```
link to message concept doc

gamarin2 commented on docs/core/transactions.md:
```diff
@@ -0,0 +1,84 @@
+Transactions messages are defined by module developers. The transaction objects themselves are SDK types that implement the [`Tx`](https://github.com/cosmos/cosmos-sdk/blob/master/types/tx_msg.go#L34-L41) interface and include an encoder and decoder:
+
+* **GetMsgs:** unwraps the transaction and returns a list of its message(s) - one transaction may have one or multiple messages.
+* **ValidateBasic:** includes lightweight, [*stateless*](../basics/tx-lifecycle.md#types-of-checks) checks used by ABCI messages [`CheckTx`](../basics/baseapp.md#checktx) and [`DeliverTx`](../basics/baseapp.md#delivertx) to validate transactions. The `StdTx` `ValidateBasic` function checks that its transactions are signed by the correct number of signers and that the fees do not exceed what the u ...
```
```suggestion
* **ValidateBasic:** includes lightweight, [*stateless*](../basics/tx-lifecycle.md#types-of-checks) checks used by ABCI messages [`CheckTx`](../basics/baseapp.md#checktx) and [`DeliverTx`](../basics/baseapp.md#delivertx) to make sure transactions are not invalid. The `StdTx` `ValidateBasic` function checks that transactions are signed by the correct number of signers and that the fees do not exceed what the user's maximum. Note that this function is distinct from the `ValidateBasic` functions for *messages*, which perform basic validity checks on messages only. For example, when [`runTx`](../basics/baseapp.md#runtx-and-runmsgs) is checking a transaction created from the [`auth`](https://github.com/cosmos/cosmos-sdk/blob/master/docs/spec/auth) module, it first runs `ValidateBasic` on each message, then runs the `auth` module AnteHandler which calls `ValidateBasic` for the transaction itself.
```

gamarin2 commented on docs/core/transactions.md:
```diff
@@ -0,0 +1,84 @@
+
+* **GetMsgs:** unwraps the transaction and returns a list of its message(s) - one transaction may have one or multiple messages.
+* **ValidateBasic:** includes lightweight, [*stateless*](../basics/tx-lifecycle.md#types-of-checks) checks used by ABCI messages [`CheckTx`](../basics/baseapp.md#checktx) and [`DeliverTx`](../basics/baseapp.md#delivertx) to validate transactions. The `StdTx` `ValidateBasic` function checks that its transactions are signed by the correct number of signers and that the fees do not exceed what the u ...
+* **TxEncoder:** Nodes running the consensus engine (e.g. Tendermint Core) are responsible for gossiping transactions and ordering them into blocks, but only handle them in the generic `[]byte` form. Transactions are always marshaled (encoded) before they are relayed to nodes, which compacts them to facilitate gossiping and helps maintain the consensus engine's separation from from application lo ...
```
link to encoding.md, not amino.md

gamarin2 commented on docs/core/transactions.md:
```diff
@@ -0,0 +1,84 @@
+* **TxEncoder:** Nodes running the consensus engine (e.g. Tendermint Core) are responsible for gossiping transactions and ordering them into blocks, but only handle them in the generic `[]byte` form. Transactions are always marshaled (encoded) before they are relayed to nodes, which compacts them to facilitate gossiping and helps maintain the consensus engine's separation from from application lo ...
+* **TxDecoder:** [ABCI](https://tendermint.com/docs/spec/abci/) calls from the consensus engine to the application, such as `CheckTx` and `DeliverTx`, are used to process transaction data to determine validity and state changes. Since transactions are passed in as `txBytes []byte`, they need to first be unmarshaled (decoded) using `TxDecoder` before any logic is applied.
+
+A transaction is created through one of the possible [interfaces](#interfaces). In the process, two contexts and an array of [messages](#messages) are created, which are then used to [generate](#transaction-generation) the transaction itself. The actual state changes triggered by transactions are enabled by the [handlers](#handlers). The rest of the document will describe each of these components ...
```
adjust links

gamarin2 commented on docs/core/transactions.md:
```diff
@@ -0,0 +1,84 @@
+
+## CLI and REST Interfaces
+
+The SDK uses several tools for building [interfaces](./interfaces.md) through which users can create transactions through the command-line or HTTP Requests. Application developers create entrypoints to the application by creating a [command-line interface](./interfaces.md#cli) or [REST interface](./interfaces.md#rest), typically found in the application's `/cmd` folder. These interfaces allow use ...
```
Remove first sentence

gamarin2 commented on docs/core/transactions.md:
```diff
@@ -0,0 +1,84 @@
+
+Note: module messages are not to be confused with [ABCI Messages](https://tendermint.com/docs/spec/abci/abci.html#messages) which define interactions between the Tendermint and application layers. While ABCI messages such as CheckTx and DeliverTx contain Transactions, which contain module Messages, they are not to be confused with the module level messages themselves
+
+The [`Msg`](https://github.com/cosmos/cosmos-sdk/blob/master/types/tx_msg.go#L10-L31) interface has five required functions.
```
No need to expand on the interface here, just link to `../building-modules/messages-and-queries.md#messages`

gamarin2 commented on docs/core/transactions.md:
```diff
@@ -0,0 +1,84 @@
+While messages contain the information for state transition logic, a transaction's other metadata and relevant information are stored in the `TxBuilder` and `CLIContext`.
+
+## Transaction Generation
+
```
Maybe start with an example command

gamarin2 commented on docs/core/transactions.md:
```diff
@@ -0,0 +1,84 @@
+
+[`Contexts`](https://godoc.org/context) are immutable objects that contain all the information needed to process a request. In the process of creating a transaction, two contexts are created: the [`CLIContext`](../interfaces/query-lifecycle.md#clicontext) and `TxBuilder`. Both are automatically generated and do not need to be defined by application developers, but do require input from the transa ...
+
+The [`TxBuilder`](https://github.com/cosmos/cosmos-sdk/blob/master/x/auth/types/txbuilder.go) contains data closely related with the processing of transactions:
```
Well it's one way of creating transaction (through `auth` module), it hsould be specified that it's not necessarily mandatory to do it this way

gamarin2 commented on docs/core/transactions.md:
```diff
@@ -0,0 +1,84 @@
+
+## Handlers
+
+The final components developers must implement to enable transactions are handlers and keeprs. Each module has a Handler to process all of the module's message types. To read more about handlers, visit the documentation for building modules [here](../building-modules/handler.md).
```
Why? Explain why they need to do that

tnachen commented:
@gamarin2 is this ready to merge?

gamarin2 commented:
I will do a final review and then merge @tnachen

gamarin2 commented on docs/core/transactions.md:
```diff
@@ -0,0 +1,80 @@
+
+## Transactions
+
+[**Transactions**](https://github.com/cosmos/cosmos-sdk/blob/master/types/tx_msg.go#L36-L43)  are objects created by end-users to trigger state changes in the application. Specifically, they are comprised of metadata held in [contexts](./context.md) and [messages](../building-modules/messages-and-queries.md) that trigger state changes within a module through the module's Handler.
```
link to handler doc

gamarin2 commented on docs/core/transactions.md:
```diff
@@ -0,0 +1,80 @@
+
+## Transaction Definition
+
+The transaction objects themselves are SDK types that implement the [`Tx`](https://github.com/cosmos/cosmos-sdk/blob/master/types/tx_msg.go#L34-L41) interface and include an encoder and decoder:
```
"implement the Tx interface, which contains the following methods:"

gamarin2 commented on docs/core/transactions.md:
```diff
@@ -0,0 +1,80 @@
+
+The transaction objects themselves are SDK types that implement the [`Tx`](https://github.com/cosmos/cosmos-sdk/blob/master/types/tx_msg.go#L34-L41) interface and include an encoder and decoder:
+
+* **GetMsgs:** unwraps the transaction and returns a list of its message(s) - one transaction may have one or multiple [messages](../building-modules/messages-and-queries.md#messages), which are defined by module developers.
```
There is a problem with the description of these 4 methods. You are conflating the description of the interface with the description of one possible implementation.

The description of the interface just explains what the expected behaviour is (e.g. `ValidateBasic` are for stateless checks, triggered when ...).

The description of one implementation explains the specific of these checks (e.g. `ValidateBasic` checks signatures and fees). 

I'm not against including one example, but it should be clearly separated. Also if you do include an example, say in what module it is implemented and link to it.

fedekunze commented:
@gzhao408 bump or someone who wants to take on this on

gamarin2 commented:
@gzhao408 any chance you can address the latest comments? 🙏

alexanderbez commented:
@gamarin2 do you mind "requesting changes"?

gzhao408 commented:
@gamarin2 done!

alexanderbez commented:
@merge when you feel this is ready @gamarin2
